src/wp-includes/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -0,0 +1,672 @@
+	 * script tag of type "application/json". Once in the browser, the state will
+	 * be parsed and used to hydrate the client-side interactivity stores and the
+	 * configuration will be available using a `getConfig` utility.
+	 *
+	 * @since 6.5.0
+	 */
+	public function print_client_interactivity_data() {
+		$store      = array();
+		$has_state  = ! empty( $this->state_data );
+		$has_config = ! empty( $this->config_data );
+
+		if ( $has_state || $has_config ) {
+			if ( $has_config ) {
+				$store['config'] = $this->config_data;
+			}
+			if ( $has_state ) {
+				$store['state'] = $this->state_data;
+			}
+			wp_print_inline_script_tag(
+				wp_json_encode(
+					$store,
+					JSON_HEX_TAG | JSON_HEX_AMP
+				),
+				array(
+					'type' => 'application/json',
+					'id'   => 'wp-interactivity-data',
+				)
+			);
+		}
+	}
+
+	/**
+	 * Registers the `@wordpress/interactivity` script modules.
+	 *
+	 * @since 6.5.0
+	 */
+	public function register_script_modules() {
+		wp_register_script_module(
+			'@wordpress/interactivity',
+			includes_url( 'js/dist/interactivity.min.js' )
+		);
+
+		wp_register_script_module(
+			'@wordpress/interactivity-router',
+			includes_url( 'js/dist/interactivity-router.min.js' ),
+			array( '@wordpress/interactivity' )
+		);
+	}
+
+	/**
+	 * Adds the necessary hooks for the Interactivity API.
+	 *
+	 * @since 6.5.0
+	 */
+	public function add_hooks() {
+		add_action( 'wp_footer', array( $this, 'print_client_interactivity_data' ) );
+	}
+
+	/**
+	 * Processes the interactivity directives contained within the HTML content
+	 * and updates the markup accordingly.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $html The HTML content to process.
+	 * @return string The processed HTML content. It returns the original content when the HTML contains unbalanced tags.
+	 */
+	public function process_directives( string $html ): string {
+		$p               = new WP_Interactivity_API_Directives_Processor( $html );
+		$tag_stack       = array();
+		$namespace_stack = array();
+		$context_stack   = array();
+		$unbalanced      = false;
+
+		$directive_processor_prefixes          = array_keys( self::$directive_processors );
+		$directive_processor_prefixes_reversed = array_reverse( $directive_processor_prefixes );
+
+		while ( $p->next_tag( array( 'tag_closers' => 'visit' ) ) && false === $unbalanced ) {
+			$tag_name = $p->get_tag();
+
+			if ( $p->is_tag_closer() ) {
+				list( $opening_tag_name, $directives_prefixes ) = end( $tag_stack );
+
+				if ( 0 === count( $tag_stack ) || $opening_tag_name !== $tag_name ) {
+
+					/*
+						* If the tag stack is empty or the matching opening tag is not the
+						* same than the closing tag, it means the HTML is unbalanced and it
+						* stops processing it.
+						*/
+					$unbalanced = true;
+					continue;
+				} else {
+
+					/*
+						* It removes the last tag from the stack.
+						*/
+					array_pop( $tag_stack );
```

Comment:
This one might not even be necessary.

```suggestion
					// Remove the last tag from the stack.
					array_pop( $tag_stack );
```